File: TinkerRocketApp/TinkerRocketApp/Models/MagCalStatus.swift
//
//  MagCalStatus.swift
//  TinkerRocketApp
//
//  Decoded form of the FC's MagCalStatusData binary frame (issue #96).
//  The wire frame is 26 bytes, little-endian, prefixed with a 0xCA byte
//  on the BLE file_ops characteristic (sibling of the 0xAA scan-results
//  prefix).  Field layout is mirrored from
//  tinkerrocket-idf/components/TR_RocketComputerTypes/RocketComputerTypes.h
//  — keep both in sync if the wire format ever changes.
//
//  Older firmware ships a 22-byte payload (no coverage_mask trailer);
//  the decoder accepts both and defaults coverageMask to 0 on old FC
//  builds so the UI degrades to timer-based prompts.
//

import Foundation

enum MagCalSubType: UInt8 {
    case idle      = 0
    case sampling  = 1
    case review    = 2
    case applied   = 3
    case aborted   = 4
    // #206 — between Accept and final NVS persist.  The chip's OFFSET
    // regs are already programmed with the new fit and the FC is
    // sampling for ~5 s while the user rotates the rocket, to verify
    // the corrected |B| stays in a tight band.  iOS shows a
    // "Verifying… rotate slowly" screen with the live |B|.
    case verifying = 5
}

enum MagCalRejectCode: UInt8 {
    case ok                       = 0
    case rTooLow                  = 1   // fitted R < 20 µT
    case rTooHigh                 = 2   // fitted R > 80 µT
    case highResidual             = 3   // RMS residual above threshold (poor sphere fit)
    case lowCoverage              = 4   // < min populated wedges (insufficient tumble coverage)
    // #206 — post-accept verification pass failed.  Originally a single
    // verifyFailed code, but the iOS message ("corrected |B| reached
    // X µT") was misleading when the gate that actually failed was
    // coverage or sample-count rather than |B| magnitude.  Split into
    // sub-codes so the UI can be specific.  In all cases the frame's
    // instantaneousFieldUT carries the worst observed |B|.
    case verifyFailed             = 5   // legacy / kept for back-compat
    case verifyTooHigh            = 6
    case verifyTooLow             = 7
    case verifyRangeWide          = 8
    case verifyLowCoverage        = 9
    case verifyFewSamples         = 10
}

struct MagCalStatus: Equatable {
    /// Sub-state of the cal flow.  Drives the iOS UI: SAMPLING shows a
    /// progress bar, REVIEW shows the fit + Accept/Retry buttons, APPLIED
    /// shows a one-shot success banner, ABORTED dismisses to settings.
    let subType: MagCalSubType

    /// Number of distinct directional wedges populated, out of 26 (3³ - 1
    /// — the (0,0,0) center cell is unreachable for a unit vector so we
    /// only count the 26 reachable wedges).  Hits ~26 on a clean tumble.
    let coverageBins: UInt8

    /// Total raw samples accumulated this run, capped at MAX_SAMPLES.
    let sampleCount: UInt16

    /// Magnitude of the most recent raw sample, in µT (post-IIS2MDC OFFSET
    /// chip subtract during SAMPLING — i.e. this is the *uncalibrated*
    /// reading the cal flow is trying to fix).  0 before the first sample.
    let instantaneousFieldUT: Float

    /// Fitted hard-iron offset in raw IIS2MDC LSB units (0.15 µT/LSB).
    /// Zero in SAMPLING/IDLE/ABORTED, populated in REVIEW/APPLIED.
    let offsetX: Int16
    let offsetY: Int16
    let offsetZ: Int16

    /// Fitted Earth-field magnitude in µT.  0 if no fit yet.  Should land
    /// in [20, 80] µT for the fit to pass the R-band sanity gate.
    let fieldR_uT: Float

    /// RMS residual of samples from the fitted sphere, in µT.  Smaller is
    /// better; values > MAG_CAL_MAX_RESIDUAL_UT (~8 µT) usually mean
    /// soft-iron distortion or a moving interferer during the tumble.
    let residualUT: Float

    /// 0 = fit accepted, non-zero = fit failed a gate; the iOS UI shows
    /// the right error and offers Retry instead of Accept.
    let rejectCode: MagCalRejectCode

    /// Bitmap of populated 3³ wedges; bit i corresponds to wedge i in
    /// directionWedge() (firmware-side TR_MagCalibrator).  Drives the
    /// per-direction progress grid and the gated orientation-prompt
    /// cycle.  Bit 13 is the unreachable centre cell and is always 0.
    /// 0 on old firmware builds that ship the 22-byte payload — UI
    /// then falls back to a simple timer cycle.
    let coverageMask: UInt32

    /// Live raw mag-vector components in µT (most recent sample).
    /// Lets the iOS UI render real-time direction feedback so the
    /// user can verify which axis is currently dominant and adjust
    /// the rocket to match the target orientation.  All zero on old
    /// firmware (< 32-byte payload).
    let liveX_uT: Float
    let liveY_uT: Float
    let liveZ_uT: Float

    /// #148 partial-coverage mask — bit i set when wedge i has 1..(N-1)
    /// samples and is not yet promoted into coverageMask.  Disjoint from
    /// coverageMask.  Lets the iOS sphere viz render 3-state cells:
    /// untouched / in-progress (this mask) / captured (coverageMask).
    /// All zero on firmware builds that don't ship the 36-byte payload —
    /// UI then falls back to 2-state (untouched / captured).
    let partialMask: UInt32

    /// Convenience: human-readable explanation for a non-zero rejectCode.
    /// Each verify-* sub-code names the specific gate that tripped so the
    /// user knows whether to re-tumble, re-rotate, or move away from
    /// interference.  instantaneousFieldUT carries the worst observed |B|.
    var rejectMessage: String {
        switch rejectCode {
        case .ok:            return "Looks good"
        case .rTooLow:       return "Field magnitude too low — interference?"
        case .rTooHigh:      return "Field magnitude too high — magnet nearby?"
        case .highResidual:  return "Sphere fit poor — likely soft-iron distortion"
        case .lowCoverage:   return "Insufficient orientation coverage — keep tumbling"
        case .verifyTooHigh:
            return String(format: "Verify failed — corrected |B| reached %.1f µT (above 70 µT cap). Try moving away from interference.",
                          instantaneousFieldUT)
        case .verifyTooLow:
            return String(format: "Verify failed — corrected |B| dropped to %.1f µT (below 20 µT floor). Cal may be over-subtracting.",
                          instantaneousFieldUT)
        case .verifyRangeWide:
            return String(format: "Verify failed — corrected |B| swung too widely (worst %.1f µT, spread > 25 µT). Hard-iron residual remained large.",
                          instantaneousFieldUT)
        case .verifyLowCoverage:
            return "Verify failed — didn't rotate the rocket through enough orientations during the verify window.  Keep rotating, then tap Done."
        case .verifyFewSamples:
            return "Verify failed — not enough samples accumulated.  Verify longer next time."
        case .verifyFailed:
            // Legacy code path — kept for back-compat with older firmware
            // builds that haven't migrated to the specific sub-codes.
            return String(format: "Verify failed — corrected |B| reached %.1f µT.",
                          instantaneousFieldUT)
        }
    }

    /// Convenience: progress fraction in [0, 1] for the SAMPLING UI.
    /// Combines the sample-count cap and the coverage gate so the bar
    /// only fills once both are healthy.
    func samplingProgress(targetSamples: UInt16, minCoverage: UInt8) -> Double {
        let s = min(Double(sampleCount) / Double(targetSamples), 1.0)
        let c = min(Double(coverageBins) / Double(minCoverage), 1.0)
        return min(s, c)
    }

    /// Decode the wire payload (bytes *after* the 0xCA discriminator).
    /// Accepts both the original 22-byte layout and the 26-byte layout
    /// that trails a uint32 coverage_mask — old firmware → coverageMask=0
    /// and UI falls back to a timer-only prompt cycle.  Returns nil if
    /// the buffer is too short to be the original payload.
    static func decode(_ bytes: [UInt8]) -> MagCalStatus? {
        guard bytes.count >= 22 else { return nil }

        // Helpers — Data slicing is alignment-safe even at odd offsets.
        func u16(_ offset: Int) -> UInt16 {
            return bytes.withUnsafeBufferPointer {
                UnsafeRawBufferPointer($0).loadUnaligned(fromByteOffset: offset, as: UInt16.self)
            }
        }
        func u32(_ offset: Int) -> UInt32 {
            return bytes.withUnsafeBufferPointer {
                UnsafeRawBufferPointer($0).loadUnaligned(fromByteOffset: offset, as: UInt32.self)
            }
        }
        func i16(_ offset: Int) -> Int16 {
            return bytes.withUnsafeBufferPointer {
                UnsafeRawBufferPointer($0).loadUnaligned(fromByteOffset: offset, as: Int16.self)
            }
        }

        // bytes[0..3]   = uint32 time_us (unused on iOS — we use phone clock for UI)
        let subTypeRaw    = bytes[4]
        let coverage      = bytes[5]
        let sampleCount   = u16(6)
        let instUTx10     = u16(8)
        let offX          = i16(10)
        let offY          = i16(12)
        let offZ          = i16(14)
        let RUTx10        = u16(16)
        let resUTx10      = u16(18)
        let rejectRaw     = bytes[20]
        // bytes[21]      = _pad
        // bytes[22..25]  = uint32 coverage_mask  (new in 26-byte payload)
        // bytes[26..31]  = int16 inst_x/y/z LSB   (new in 32-byte payload)
        // bytes[32..35]  = uint32 partial_mask   (new in 36-byte payload, #148)
        let coverageMask: UInt32 = (bytes.count >= 26) ? u32(22) : 0
        let liveX_lsb: Int16 = (bytes.count >= 32) ? i16(26) : 0
        let liveY_lsb: Int16 = (bytes.count >= 32) ? i16(28) : 0
        let liveZ_lsb: Int16 = (bytes.count >= 32) ? i16(30) : 0
        let partialMask: UInt32 = (bytes.count >= 36) ? u32(32) : 0

        // IIS2MDC sensitivity is 0.15 µT/LSB (datasheet 9.13).  Mirrors
        // the firmware-side IIS2MDC_LSB_TO_uT constant.
        let LSB_TO_uT: Float = 0.15

        let sub    = MagCalSubType(rawValue: subTypeRaw) ?? .idle
        let reject = MagCalRejectCode(rawValue: rejectRaw) ?? .ok

        return MagCalStatus(
            subType: sub,
            coverageBins: coverage,
            sampleCount: sampleCount,
            instantaneousFieldUT: Float(instUTx10) / 10.0,
            offsetX: offX,
            offsetY: offY,
            offsetZ: offZ,
            fieldR_uT: Float(RUTx10) / 10.0,
            residualUT: Float(resUTx10) / 10.0,
            rejectCode: reject,
            coverageMask: coverageMask,
            liveX_uT: Float(liveX_lsb) * LSB_TO_uT,
            liveY_uT: Float(liveY_lsb) * LSB_TO_uT,
            liveZ_uT: Float(liveZ_lsb) * LSB_TO_uT,
            partialMask: partialMask
        )
    }
}

// Six cardinal-axis "wedge caps" that the UI uses to drive the gated
// prompt cycle (issue #96 follow-up).  These are the firmware
// directionWedge() indices for unit vectors near ±X / ±Y / ±Z — see the
// 3×3×3 encoding in TR_MagCalibrator.cpp (bx*9 + by*3 + bz with each
// component in {0,1,2} for {<-T, [-T,T], >T}).
//
// Body-frame convention (empirical, verified by the user on the flight
// computer): nose-up populates the +X wedge.  Sides and faces follow
// from the right-hand rule around that — if any axis turns out to be
// flipped on the rocket, swap the +Y/+Z assignments here without
// touching the firmware.
enum MagCalAxis: CaseIterable {
    case noseUp, noseDown
    case rightSide, leftSide
    case frontFace, backFace

    /// Bit index inside MagCalStatus.coverageMask.
    var wedgeBit: UInt32 {
        switch self {
        // bx, by, bz ∈ {0,1,2} → index = bx*9 + by*3 + bz
        case .noseUp:    return 2 * 9 + 1 * 3 + 1  // (+,0,0) = 22  — user-confirmed: Up = +X
        case .noseDown:  return 0 * 9 + 1 * 3 + 1  // (-,0,0) =  4
        case .rightSide: return 1 * 9 + 2 * 3 + 1  // (0,+,0) = 16
        case .leftSide:  return 1 * 9 + 0 * 3 + 1  // (0,-,0) = 10
        case .frontFace: return 1 * 9 + 1 * 3 + 2  // (0,0,+) = 14
        case .backFace:  return 1 * 9 + 1 * 3 + 0  // (0,0,-) = 12
        }
    }

    /// Big arrow glyph for the prompt card.
    var icon: String {
        switch self {
        case .noseUp:    return "arrow.up"
        case .noseDown:  return "arrow.down"
        case .rightSide: return "arrow.right"
        case .leftSide:  return "arrow.left"
        case .frontFace: return "arrow.up.right"
        case .backFace:  return "arrow.down.left"
        }
    }

    /// Imperative direction for the prompt card.
    var prompt: String {
        switch self {
        case .noseUp:    return "Point the nose UP"
        case .noseDown:  return "Point the nose DOWN"
        case .rightSide: return "Lay it on its RIGHT side"
        case .leftSide:  return "Lay it on its LEFT side"
        case .frontFace: return "Tilt the FRONT face up"
        case .backFace:  return "Tilt the BACK face up"
        }
    }

    /// Short label for the per-direction status grid.
    var shortLabel: String {
        switch self {
        case .noseUp:    return "Up"
        case .noseDown:  return "Down"
        case .rightSide: return "Right"
        case .leftSide:  return "Left"
        case .frontFace: return "Front"
        case .backFace:  return "Back"
        }
    }

    /// Which body-frame axis (X/Y/Z) this orientation targets.
    var component: MagCalComponent {
        switch self {
        case .noseUp, .noseDown:     return .x
        case .rightSide, .leftSide:  return .y
        case .frontFace, .backFace:  return .z
        }
    }

    /// Sign the target component should have when correctly oriented.
    /// e.g. nose-up → +X (sign +1); nose-down → -X (sign -1).
    var sign: Float {
        switch self {
        case .noseUp, .rightSide, .frontFace: return  1.0
        case .noseDown, .leftSide, .backFace: return -1.0
        }
    }
}

/// Sensor-frame component axis labels for the direction-feedback bars.
enum MagCalComponent: String, CaseIterable {
    case x = "X"
    case y = "Y"
    case z = "Z"

    /// Pull the live µT reading for this component out of a status frame.
    func value(in status: MagCalStatus) -> Float {
        switch self {
        case .x: return status.liveX_uT
        case .y: return status.liveY_uT
        case .z: return status.liveZ_uT
        }
    }
}

extension MagCalStatus {
    /// True iff the wedge for `axis` has been populated in this run.
    func isAxisCovered(_ axis: MagCalAxis) -> Bool {
        return (coverageMask & (UInt32(1) << axis.wedgeBit)) != 0
    }

    /// Magnitude of the fitted hard-iron offset vector, in µT.
    /// (#207) Surfaced on the Review screen so the user can tell at a
    /// glance whether the cal converged on a center close to the sensor
    /// origin (good) or absorbed an external interferer (bad).
    var centerMagnitudeUT: Float {
        let x = Float(offsetX), y = Float(offsetY), z = Float(offsetZ)
        // IIS2MDC: 0.15 µT/LSB.  Mirrors LSB_TO_uT in decode().
        return 0.15 * (x*x + y*y + z*z).squareRoot()
    }

    /// |c| / R as a fraction in [0, ∞).  Ratios approaching 1 mean the
    /// applied offset is the same magnitude as Earth's field — rotating
    /// the rocket then sweeps |B| across roughly [R - |c|, R + |c|],
    /// which can clip the EKF's [15, 80] µT mag input gate at the
    /// extremes.  Zero when no fit has converged yet (R == 0).
    var centerToRRatio: Float {
        guard fieldR_uT > 0 else { return 0 }
        return centerMagnitudeUT / fieldR_uT
    }

    /// UI severity bucket for the *absolute* center magnitude |c|.
    /// Only flags an actual problem — a "large" bias in itself isn't
    /// bad (the cal absorbs it), the only failure mode for |c| alone is
    /// approaching the chip's ±4915 µT OFFSET register range, beyond
    /// which the IIS2MDC can't fully subtract the bias.  Typical fresh
    /// new-PCB IIS2MDC sits around 1.7 mT — well under 4500 µT, so the
    /// row stays green.
    enum CenterWarning {
        case ok       // |c| < 4500 µT (chip can subtract it)
        case high     // |c| ≥ 4500 µT (approaching chip OFFSET range)
    }

    var centerWarning: CenterWarning {
        return centerMagnitudeUT >= 4500 ? .high : .ok
    }
}

// Match firmware-side gates (RocketComputerTypes.h) so the UI doesn't
// drift from the FC's accept/reject decision.  The FC is the source of
// truth via rejectCode — these constants are just for the progress UI.
enum MagCalConstants {
    static let maxSamples: UInt16 = 3200   // 32 accel-wedges × 100 slots each (#148)
    static let minSamples: UInt16 = 500
    static let minCoverageBins: UInt8 = 22 // MAG_CAL_MIN_COVERAGE_BINS, #148 (22/32)

    // #148 — verify-window gates.  Keep in sync with
    // MAG_CAL_VERIFY_* in RocketComputerTypes.h.  iOS uses these to
    // show real-time pass/fail feedback while the user rotates during
    // verify — the firmware applies the same thresholds when the user
    // taps Done (or the 60 s safety timeout fires).
    static let verifyMinUT: Float        = 20.0
    static let verifyMaxUT: Float        = 70.0
    static let verifyRangeUT: Float      = 25.0
    static let verifyMinCoverageBins: UInt8 = 8
    static let verifyMinSamples: UInt16  = 100
}

File: TinkerRocketAndroid/core/protocol/src/main/kotlin/com/tinkerbug/tinkerrocket/protocol/MagCalStatus.kt
package com.tinkerbug.tinkerrocket.protocol

import kotlin.math.sqrt

/** Sub-state of the mag-cal flow (MagCalStatusData.sub_type). */
public enum class MagCalSubType(public val raw: Int) {
    IDLE(0),
    SAMPLING(1),
    REVIEW(2),
    APPLIED(3),
    ABORTED(4),

    /**
     * #206 — between Accept and final NVS persist: chip OFFSET regs already
     * programmed, FC sampling ~5 s while the user rotates to verify the
     * corrected |B| stays in a tight band.
     */
    VERIFYING(5);

    /**
     * Is a calibration in flight, such that leaving the screen must abort it?
     *
     * The set matters in both directions.  Leaving during any of these three
     * strands the FC in MAG_CALIBRATION with the magnetometer offsets zeroed,
     * and that state has NO launch-detect failsafe by design — kinematicChecks
     * is skipped in MAG_CALIBRATION, so nothing ever clears it on its own.
     * VERIFYING is the sharpest of the three: a 60 s firmware timeout runs
     * evaluateVerify() whether or not the app is still watching, so walking
     * away can commit a permanent NVS calibration nobody reviewed.
     *
     * APPLIED must NOT be in the set. The FC rests in APPLIED whenever a
     * calibration exists, and MagCalibrator::abort() forces ABORTED from ANY
     * state — so aborting on the way out of a successful save would throw the
     * calibration away.
     */
    public val needsAbortOnLeave: Boolean
        get() = this == SAMPLING || this == REVIEW || this == VERIFYING

    public companion object {
        /** Unknown raw values fall back to [IDLE], mirroring the iOS `?? .idle`. */
        public fun fromRaw(raw: Int): MagCalSubType =
            entries.firstOrNull { it.raw == raw } ?: IDLE
    }
}

/** Fit accept/reject gate result (MagCalStatusData.reject_code). */
public enum class MagCalRejectCode(public val raw: Int) {
    OK(0),
    R_TOO_LOW(1),             // fitted R < 20 µT
    R_TOO_HIGH(2),            // fitted R > 80 µT
    HIGH_RESIDUAL(3),         // RMS residual above threshold (poor sphere fit)
    LOW_COVERAGE(4),          // < min populated wedges (insufficient tumble coverage)

    // #206 — post-accept verification sub-codes; the frame's
    // instantaneousFieldUt carries the worst observed |B| in all cases.
    VERIFY_FAILED(5),         // legacy / kept for back-compat
    VERIFY_TOO_HIGH(6),
    VERIFY_TOO_LOW(7),
    VERIFY_RANGE_WIDE(8),
    VERIFY_LOW_COVERAGE(9),
    VERIFY_FEW_SAMPLES(10);

    public companion object {
        /** Unknown raw values fall back to [OK], mirroring the iOS `?? .ok`. */
        public fun fromRaw(raw: Int): MagCalRejectCode =
            entries.firstOrNull { it.raw == raw } ?: OK
    }
}

/**
 * MagCalStatus — decoded form of the FC's MagCalStatusData binary frame
 * (issue #96), arriving on the BLE file_ops characteristic behind a 0xCA
 * discriminator.  Current wire payload is 36 bytes LE; the format grew over
 * time (22-byte → +coverage_mask 26 → +live LSBs 32 → +partial_mask 36,
 * #148) and [decode] accepts the shorter legacy payloads, defaulting the
 * missing trailers to 0 so older FC builds degrade to timer-based prompts.
 *
 * Field layout mirrors MagCalStatusData in RocketComputerTypes.h
 * (static_assert sizeof == 36).  The wire time_us (bytes 0..3) is not
 * exposed — iOS uses the phone clock for UI, and this port pins iOS.
 */
public data class MagCalStatus(
    /** Sub-state of the cal flow (drives the client UI). */
    val subType: MagCalSubType,

    /** Distinct directional wedges populated, out of 32 (#148). */
    val coverageBins: Int,      // u8

    /** Total raw samples accumulated this run, capped at MAX_SAMPLES. */
    val sampleCount: Int,       // u16

    /**
     * Magnitude of the most recent raw sample, in µT (post-OFFSET subtract
     * during SAMPLING — the *uncalibrated* reading).  0 before first sample.
     */
    val instantaneousFieldUt: Float,   // wire u16 ×10

    /**
     * Fitted hard-iron offset in raw IIS2MDC LSB units (0.15 µT/LSB).
     * Zero in SAMPLING/IDLE/ABORTED, populated in REVIEW/APPLIED.
     */
    val offsetX: Int,           // i16
    val offsetY: Int,
    val offsetZ: Int,

    /** Fitted Earth-field magnitude in µT; should land in [20, 80]. */
    val fieldRUt: Float,        // wire u16 ×10

    /** RMS residual of samples from the fitted sphere, in µT. */
    val residualUt: Float,      // wire u16 ×10

    /** 0 = fit accepted, non-zero = fit failed a gate. */
    val rejectCode: MagCalRejectCode,

    /**
     * Bitmap of populated 3³ wedges; bit 13 (unreachable centre cell) is
     * always 0.  0 on old firmware that ships the 22-byte payload.
     */
    val coverageMask: Long,     // u32

    /** Live raw mag-vector components in µT; all 0 on < 32-byte payloads. */
    val liveXUt: Float,
    val liveYUt: Float,
    val liveZUt: Float,

    /**
     * #148 partial-coverage mask — wedges with 1..(N-1) samples, disjoint
     * from [coverageMask].  0 on < 36-byte payloads.
     */
    val partialMask: Long,      // u32
) {
    /**
     * Magnitude of the fitted hard-iron offset vector, in µT (#207).
     * IIS2MDC: 0.15 µT/LSB, mirrors LSB_TO_UT in [decode].
     */
    public val centerMagnitudeUt: Float
        get() {
            val x = offsetX.toFloat(); val y = offsetY.toFloat(); val z = offsetZ.toFloat()
            return 0.15f * sqrt(x * x + y * y + z * z)
        }

    /**
     * |c| / R as a fraction in [0, ∞).  Ratios approaching 1 mean the
     * applied offset rivals Earth's field — rotating then sweeps |B| across
     * roughly [R − |c|, R + |c|], which can clip the EKF's [15, 80] µT mag
     * input gate.  Zero when no fit has converged yet (R == 0).
     */
    public val centerToRRatio: Float
        get() = if (fieldRUt > 0) centerMagnitudeUt / fieldRUt else 0f

    /**
     * UI severity bucket for the *absolute* center magnitude |c|.  The only
     * failure mode for |c| alone is approaching the chip's ±4915 µT OFFSET
     * register range, beyond which the IIS2MDC can't fully subtract the
     * bias.  Typical fresh new-PCB IIS2MDC sits around 1.7 mT — well under
     * 4500 µT, so the row stays [OK].
     */
    public enum class CenterWarning {
        OK,       // |c| < 4500 µT (chip can subtract it)
        HIGH,     // |c| ≥ 4500 µT (approaching chip OFFSET range)
    }

    public val centerWarning: CenterWarning
        get() = if (centerMagnitudeUt >= 4500f) CenterWarning.HIGH else CenterWarning.OK

    public companion object {
        /** Original (oldest accepted) payload size. */
        public const val MIN_SIZE: Int = 22

        /** IIS2MDC sensitivity, 0.15 µT/LSB (datasheet 9.13); mirrors firmware IIS2MDC_LSB_TO_uT. */
        public const val LSB_TO_UT: Float = 0.15f

        /**
         * Decode the wire payload (bytes *after* the 0xCA discriminator).
         * Accepts the legacy 22-byte layout, the +coverage_mask layout, the
         * +live-LSB layout, and the current 36-byte layout; missing trailers
         * default to 0.  Returns null if the buffer is too short to be even
         * the original 22-byte payload.
         */
        public fun decode(payload: ByteArray): MagCalStatus? {
            if (payload.size < MIN_SIZE) return null
            val b = LeBuffer(payload)

            b.u32()                       // bytes[0..3] time_us — unused (phone clock drives UI)
            val subTypeRaw = b.u8()       // [4]
            val coverage = b.u8()         // [5]
            val sampleCount = b.u16()     // [6..7]
            val instUtX10 = b.u16()       // [8..9]
            val offX = b.i16()            // [10..11]
            val offY = b.i16()            // [12..13]
            val offZ = b.i16()            // [14..15]
            val rUtX10 = b.u16()          // [16..17]
            val resUtX10 = b.u16()        // [18..19]
            val rejectRaw = b.u8()        // [20]
            // [21] = _pad
            // [22..25] u32 coverage_mask  (new in 26-byte payload)
            // [26..31] i16 inst_x/y/z LSB (new in 32-byte payload)
            // [32..35] u32 partial_mask   (new in 36-byte payload, #148)
            val coverageMask = if (payload.size >= 26) LeBuffer(payload, 22).u32() else 0L
            val liveXLsb = if (payload.size >= 32) LeBuffer(payload, 26).i16() else 0
            val liveYLsb = if (payload.size >= 32) LeBuffer(payload, 28).i16() else 0
            val liveZLsb = if (payload.size >= 32) LeBuffer(payload, 30).i16() else 0
            val partialMask = if (payload.size >= 36) LeBuffer(payload, 32).u32() else 0L

            return MagCalStatus(
                subType = MagCalSubType.fromRaw(subTypeRaw),
                coverageBins = coverage,
                sampleCount = sampleCount,
                instantaneousFieldUt = instUtX10 / 10.0f,
                offsetX = offX,
                offsetY = offY,
                offsetZ = offZ,
                fieldRUt = rUtX10 / 10.0f,
                residualUt = resUtX10 / 10.0f,
                rejectCode = MagCalRejectCode.fromRaw(rejectRaw),
                coverageMask = coverageMask,
                liveXUt = liveXLsb * LSB_TO_UT,
                liveYUt = liveYLsb * LSB_TO_UT,
                liveZUt = liveZLsb * LSB_TO_UT,
                partialMask = partialMask,
            )
        }
    }
}
